src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ShardStateCollector.java

```diff
@@ -0,0 +1,132 @@
+package com.amazon.opendistro.elasticsearch.performanceanalyzer.collectors;
+
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.ESResources;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.AllMetrics;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.MetricsConfiguration;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.MetricsProcessor;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.jooq.tools.StringUtils;
+
+import static com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics.SHARD_PRIMARY;
+import static com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics.SHARD_REPLICA;
+
+public class ShardStateCollector extends PerformanceAnalyzerMetricsCollector implements MetricsProcessor {
+    public static final int SAMPLING_TIME_INTERVAL = MetricsConfiguration.CONFIG_MAP.get(ShardStateCollector.class).samplingInterval;
+    private static final Logger LOG = LogManager.getLogger(ShardStateCollector.class);
+    private static final int KEYS_PATH_LENGTH = 0;
+    private StringBuilder value;
+
+    public ShardStateCollector() {
+        super(SAMPLING_TIME_INTERVAL, "ShardsStateCollector");
+        value = new StringBuilder();
+    }
+
+    @Override
+    void collectMetrics( long startTime) {
+        if (ESResources.INSTANCE.getClusterService() == null) {
+            return;
+        }
+        ClusterState clusterState = ESResources.INSTANCE.getClusterService().state();
+
+        try {
+            value.setLength(0);
+            value.append(PerformanceAnalyzerMetrics.getJsonCurrentMilliSeconds())
+                    .append(PerformanceAnalyzerMetrics.sMetricNewLineDelimitor);
+            for(ShardRouting shard : clusterState.routingTable().allShards()) {
```

Comment:
Because we are scanning through all shards in the cluster and because this will be done on all nodes, that's a lot of redundant information. Maybe we want to filter by shards resident on this node alone ?

Reply:
If we start publishing shard info local to each node, we will loose the shard information if a node leaves the cluster which we are getting from global cluster state. This way we would never know of the unassigned shards in the cluster.

Reply:
For shards resident on nodes that are no longer part of the cluster, can the current master gather that data ? WE should try to trim down the set that each node emits otherwise, this collector will be slower and we will balloon on disk space usage and if there are metrics collectors sitting outside and consuming the metrics, they will send a lot of redundant bytes over the wire.
Let's think of ways to reduce the data generation here. I am afraid, it might be too much for some clusters.

Reply:
As discussed, we will not publish the stats for active shard. This will considerably reduce the consumed disk space